### PR TITLE
Log warn when mail not sent

### PR DESCRIPTION
**Pull request by MiguelAHM**
Depends on #1404

### CODE REVIEW

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java [hunk 1]
```diff
@@ -112,20 +112,15 @@ private boolean notifyOriginAutnum(final PreparedUpdate update) {
...
-    private void addNotifications(final Map<CIString, Notification> notifications, final PreparedUpdate update, final UpdateContext updateContext) {
+    public void addNotifications(final Map<CIString, Notification> notifications, final PreparedUpdate update,
+                            final UpdateContext updateContext) {
         final RpslObject object = update.getReferenceObject();
 
         switch (updateContext.getStatus(update)) {
             case SUCCESS:
                 if (updateContext.getAction(update) != Action.NOOP) {
                     addVersionId(update, updateContext);
-                    add(notifications, updateContext, update, Notification.Type.SUCCESS, Collections.singletonList(object), AttributeType.NOTIFY);
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java:122]: Why this change is required? We already adding versionId just above

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/common/NotificationValidator.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        this.emailStatusDao = emailStatusDao;
+        this.updateNotifier = updateNotifier;
+    }
+
+    @Override
+    public List<Message> performValidation(final PreparedUpdate update, final UpdateContext updateContext) {
+        final PreparedUpdate preparedUpdate = updateContext.getPreparedUpdate(update);
+        final Map<CIString, Notification> notifications = Maps.newHashMap();
+        if (preparedUpdate != null && !updateNotifier.notificationsDisabledByOverride(preparedUpdate)) {
+            updateNotifier.addNotificationsWithoutVersioning(notifications, preparedUpdate, updateContext, update.getReferenceObject());
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/common/NotificationValidator.java:46]: This validator should add warnings only if notificationsDisabledByOverride is false. So it should return from here and not do anything

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceTestIntegration.java [hunk 3]
```diff
@@ -5750,6 +5767,59 @@ public void update_route6_in_region_no_redirect() {
 
     }
 
+    @Test
+    public void unsubscribed_notify_user_gets_warn_when_updating() {
+        final String unsubscribedEmail = "test@ripe.net";
```
Comment (maggarwal13) [inline whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceTestIntegration.java:5772]: Can you either update existing or create new one where atleast one  notify email where it is not  unsubscribed or undelieverd and thats why no warning for that email address

File: whois-commons/src/main/java/net/ripe/db/whois/common/dao/EmailStatusDao.java [hunk 4]
```diff
@@ -26,6 +32,19 @@ public void createEmailStatus(final String email, final EmailStatus emailStatus)
                 LocalDateTime.now());
     }
 
+    public Map<String, String> getEmailStatus(final Set<String> emailAddresses) {
+        Map<String, String> results = Maps.newHashMap();
+
+        namedParameterJdbcTemplate.query(
+                "SELECT email, status from email_status where email in (:emails)",
```
Comment (maggarwal13) [inline whois-commons/src/main/java/net/ripe/db/whois/common/dao/EmailStatusDao.java:39]: Correct formatting? FROM, IN and WHERE should also be in caps?

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java [hunk 5]
```diff
@@ -101,7 +101,7 @@ private void addVersionId(final PreparedUpdate preparedUpdate, final UpdateConte
         }
     }
 
-    private boolean notificationsDisabledByOverride(final PreparedUpdate preparedUpdate) {
+    public boolean notificationsDisabledByOverride(final PreparedUpdate preparedUpdate) {
         final OverrideOptions overrideOptions = preparedUpdate.getOverrideOptions();
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java:105]: Can you rename to shouldNotifyUser ? and use negation afterwards?

File: whois-update/src/main/java/net/ripe/db/whois/update/domain/UpdateMessages.java [hunk 6]
```diff
@@ -616,6 +616,10 @@ public static Message statusCannotBeRemoved() {
         return new Message(Type.WARNING, "\"status:\" attribute cannot be removed");
     }
 
+    public static Message emailCanNotBeSent(final String email, final String emailStatus) {
+        return new Message(Type.WARNING, "Not sending notification to %s because it is %s.", email, emailStatus);
```
[the remark this answers is not in the document]
Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/domain/UpdateMessages.java:620]: I also think the error message should be short? Something like <email> is unsubscribed or <email> is undeliverable else it can become a really long warnings ? But feel free to Ignore if you think otherwise

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java (hunk 5 above)
Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java:105]: Sorry it was like this before so lets leave it as it is.

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java [hunk 7]
```diff
@@ -112,20 +112,15 @@ private boolean notifyOriginAutnum(final PreparedUpdate update) {
...
-                    add(notifications, updateContext, update, Notification.Type.SUCCESS, rpslObjectDao.getByKeys(ObjectType.MNTNER, object.getValuesForAttribute(AttributeType.MNT_BY)), AttributeType.MNT_NFY);
-                    add(notifications, updateContext, update, Notification.Type.SUCCESS_REFERENCE, rpslObjectDao.getByKeys(ObjectType.ORGANISATION, update.getDifferences(AttributeType.ORG)), AttributeType.REF_NFY);
-                    add(notifications, updateContext, update, Notification.Type.SUCCESS_REFERENCE, rpslObjectDao.getByKeys(ObjectType.IRT, update.getDifferences(AttributeType.MNT_IRT)), AttributeType.IRT_NFY);
-                    if (notifyOriginAutnum(update)) {
-                        add(notifications, updateContext, update, Notification.Type.SUCCESS_REFERENCE, rpslObjectDao.getByKeys(ObjectType.AUT_NUM, update.getDifferences(AttributeType.ORIGIN)), AttributeType.NOTIFY);
-                    }
+                    addNotificationsWithoutVersioning(notifications, update, updateContext, object);
                 }
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java:124]: This needs to be done for both success and failure

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/common/NotificationValidator.java (hunk 2 above)
Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/common/NotificationValidator.java]: Revert the condtion logic was correct

File: whois-update/src/main/java/net/ripe/db/whois/update/domain/UpdateMessages.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-update/src/main/java/net/ripe/db/whois/update/domain/UpdateMessages.java]: I am fine putting EmailStatus in lowercase instead of uppercase.

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceTestIntegration.java (hunk 3 above)
Comment (MiguelAHM) [reply whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceTestIntegration.java:5772]: we have a bunch of test that are already doing that, right? updates that are checking the warnings already in this warning doesn't exist among them

File: whois-commons/src/main/java/net/ripe/db/whois/common/dao/EmailStatusDao.java (hunk 4 above)
Comment (MiguelAHM) [reply whois-commons/src/main/java/net/ripe/db/whois/common/dao/EmailStatusDao.java]: it should be in uppercase, because they are keywords, you are right.
In the code we have a mix.

File: whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java (hunk 7 above)
Comment (MiguelAHM) [reply whois-update/src/main/java/net/ripe/db/whois/update/handler/UpdateNotifier.java:124]: should be already working like this, right?
for example when there is an error related to auth we are sending an email to upd-to user

**Comment by MiguelAHM** [comment]
test ok
https://gitlab.ripe.net/swe-database-team/whois/-/pipelines/533832

**Review by maggarwal13 — APPROVED** [review]
